ion-schema/src/isl/isl_type_reference.rs

```diff
@@ -66,53 +86,31 @@ impl IslTypeRef {
     }
 }
 
+//TODO: can add occurs here to store occurs range information for variably occurring type
```

Comment:
I recommend that we keep `occurs` and nullability separate from each other. It's better to structure it like this: 
```
struct VariableOccuringTypeRef {
    occurs: SomeRange,
    type: RegularTypeRefThatCouldBeNullable
}
```
We don't want to have any confusion about the precedence between `Nullable` and `occurs`.

Here's an example to help illustrate what I mean.
The following is illegal in ISL 2.0, and for good reason. Should `$null_or` count towards the `occurs` requirement?
```
// ...
fields: {
  foo: $null_or::{ type: string, occurs: 2 }
}
// ...
```

If yes, then we should just write `{ type: $null_or::string, occurs: 2 }`. If `null` does not count towards the occurs limit, then we have a mess of surprising behavior on our hands. (Does it mean two strings plus any number of `null`? Maybe `$null_or` has the default occurs. Does it mean optional `null` _xor_ two strings? Or maybe it means we can have optional `null` _and_ two strings?)

Unless we have a very compelling performance reason, I think we're better off structuring the data in Rust in a way that helps facilitate correct understanding of the data.

Reply:
Modified with: [adds implementation of VariablyOccurringTypeRef](https://github.com/amazon-ion/ion-schema-rust/pull/165/commits/4f10fe0d0fd74ef921f95e029475edaaa09438d0)